clippy_utils/src/ty/mod.rs

```diff
@@ -1376,3 +1376,44 @@ pub fn option_arg_ty<'tcx>(cx: &LateContext<'tcx>, ty: Ty<'tcx>) -> Option<Ty<'t
         _ => None,
     }
 }
+
+/// Check if a Ty<'_> of `Iterator` has side effects when iterated over by checking if it
+/// captures any mutable references or equivalents.
+pub fn is_iter_with_side_effects<'tcx>(cx: &LateContext<'tcx>, iter_ty: Ty<'tcx>) -> bool {
+    let mut phantoms = FxHashSet::default();
+    has_non_owning_mutable_access(cx, &mut phantoms, iter_ty)
+}
+
+/// Check if `ty` contains mutable references or equivalent, which includes:
+/// - A mutable reference/pointer.
+/// - A reference/pointer to a non-`Freeze` type.
+/// - A `PhantomData` type containing any of the previous.
+fn has_non_owning_mutable_access<'tcx>(
+    cx: &LateContext<'tcx>,
+    phantoms: &mut FxHashSet<Ty<'tcx>>,
+    ty: Ty<'tcx>,
+) -> bool {
+    match ty.kind() {
+        ty::Adt(adt_def, args) if adt_def.is_phantom_data() => {
+            phantoms.insert(ty)
```

Comment:
If you normalize types then you'll need to insert every normalized type seen, otherwise this could still recurse infinitely.